src/Core/Element.js

```diff
@@ -416,6 +400,61 @@ function getElementModule (element) {
 	return createElementFragment(getElementDefinition(element))
 }
 
+/**
+ * @param {any} type
+ * @return {number}
+ */
+function getElementIdentity (type) {
+	switch (typeof type) {
+		case 'string':
+			return SharedElementNode
+		case 'function':
+			return SharedElementComponent
+		case 'number':
+		case 'symbol':
+			return type === SymbolForFragment ? SharedElementFragment : SharedElementNode
+		default:
+			return thenable(type) ? SharedElementPromise : isValidElement(type) ? -type.id : SharedElementNode
+	}
+}
+
+/**
+ * @param {object} props
+ * @return {boolean}
+ */
+function isValidProps (props) {
+	if (typeof props === 'object' && props[SymbolForIterator] === undefined)
+		switch (props.constructor) {
+			default:
+				if (ArrayIsArray(props))
```

Comment:
can't this be moved up into the `if` above? moreover this whole switch statement can be moved into the `if`

Reply:
The switch block has been specifically written like this due to the nature of the optimization i was trying to archive with it.

The default branch doesn't appear last because there are cases(Objects from different realms) when the `Object` clause is not satisfied but at the same time it is a valid props-like `Object`. 

The only other object we want to special case in that case is Arrays, but then again that is already handled by the iterator lookup in the if block(a reason i also stuck the iterator symbol to the `Element` and `List` prototypes is to ). 

So the above default block is for very niche condition one of which is an `Object` from another realm and another is when there is an Array from another realm and specifically in environments(IE9) that do not support the iterator protocol. Everything else falls through to `Object`.

This and `createElement` are without a doubt the hottest functions(according to a JIT) in the lib. So any ideas improve this would have far reaching perf benefits.

Reply:
Fascinating, re-reading it this optimization is quite clever.
I now see the intent is to skip `ArrayIsArray` in the average case (`props.constructor === Object`).

As for performance , `value == null` (on `createElement()`) will force a slow-path because it defaults to `{}` (also sets `i = 2`, which should presumably be `i = 1`).
Simply move the `value == null` check within `isValidProps`

```js
function createElement (type, value) {
	var i = isValidProps(value) ? 2 : 1
	var props = i === 2 ? value : {}
```